torch/distributed/_composable/replicate.py

```diff
@@ -1,15 +1,17 @@
-from typing import List, Tuple
+from typing import Iterable, Optional, Set, Tuple
 
 import torch
 import torch.nn as nn
 
 from torch.nn.parallel import DistributedDataParallel
-from .contract import _get_registry, contract
+
+from .contract import contract
 
 
 @contract()
 def replicate(
     module: nn.Module,  # NOTE: contract now supports single module only
+    ignored_modules: Optional[Iterable[torch.nn.Module]] = None,
```

Comment:
curious, can trec_shard add entry to registry on submodules it swapped, so that replicate can detect automatically?

Reply:
Agree, and that's the spirit of these *composable* APIs, `ignore_modules` may look like a temporary workaround solution.

Reply:
This is indeed the right way to do so. currently trec_shard adds an entry for the top level, but does not do so for individually sharded submodules. 

This is also why fully_shard needs ignored_modules today. For this PR, I think it'd be best to unblock by adding ignored_module support, and then follow up with trec_shard to fix the registry and remove ignored_modules.

Also, if user calls replicate() first and then trec_shard, we would still need `ignored_modules`, right? Unless we add a policy or similar to replicate API.